File: 141-LinkedListCycle.cs
//141. Linked List Cycle
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public bool HasCycle(ListNode head) {
        HashSet<ListNode> nodesSeen = new HashSet<ListNode>();
        
        while (head != null) {
            if (nodesSeen.Contains(head)) {
                return true;
            }
            nodesSeen.Add(head);
            head = head.next;
        }
        return false;
    }
    
        
    
}

File: LeetCode Java/Data Structure I/141-LinkedListCycle.java
/*
141. Linked List Cycle

Given head, the head of a linked list, determine if the linked list has a cycle in it.
There is a cycle in a linked list if there is some node in the list that can be reached again by continuously
following the next pointer. Internally, pos is used to denote the index of the node that tail's next pointer is connected to. Note that pos is not passed as a parameter.
Return true if there is a cycle in the linked list. Otherwise, return false.
Example:
    Input: head = [3,2,0,-4], pos = 1
    Output: true
    Explanation: There is a cycle in the linked list, where the tail connects to the 1st node (0-indexed).
*/
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    //Time complexity :O(n). Let us denote n as the total number of nodes in the linked list
    //space complexity : O(1). We only use two nodes (slow and fast) 
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }

        ListNode slow = head;
        ListNode fast = head.next;
        //if fast == slow, fast has went all the way around the loop, therefore it exists, break
        while (slow != fast) { 
            if (fast == null || fast.next == null) {
                return false;
            }
            slow = slow.next;
            fast = fast.next.next; //double speed
        }
        return true;
    }
}